### DISCUSSION TITLE
@Dependency property wrapper does not compile inside EntityQuery conforming struct

### DISCUSSION THREAD

**Discussion by twocentstudios** [pointfreeco/swift-dependencies#230] [General]
### Description

I'm trying to use the `@Dependency` property wrapper within a struct conforming to the `EntityQuery` protocol in the `AppIntents` framework. Currently I'm trying to define an `EntityQuery` in a widget extension.

I'm getting a compiler error on the `@Dependency` line in the struct body and within the `init` method: `Type of expression is ambiguous without a type annotation`.

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-dependencies/issues) or [discussion](https://github.com/pointfreeco/swift-dependencies/discussions).

### Expected behavior

The target compiles and I can use the resolved dependency.

### Actual behavior

Compiler error: `Type of expression is ambiguous without a type annotation` (see code below)

### Steps to reproduce

- Create new SwiftUI project
- Add a WidgetExtension
- Paste the following in the generated `Widget.swift` file

```swift
import AppIntents
import Dependencies

// No error; a struct non-conforming to EntityQuery compiles
public struct SimpleModel {
    @Dependency(\.exampleDependency) var db
}

// Setting up a minimal `AppEntity`/`EntityQuery` that uses an `ExampleDependency`
@DependencyClient struct ExampleDependency { }
extension ExampleDependency: DependencyKey {
    static var liveValue: ExampleDependency { .init() }
}
extension DependencyValues {
    var exampleDependency: ExampleDependency {
        get { self[ExampleDependency.self] }
        set { self[ExampleDependency.self] = newValue }
    }
}

public struct AppEntityConformingModel: AppEntity {
    public static var defaultQuery: AppEntityConformingModelQuery { AppEntityConformingModelQuery() }
    public static var typeDisplayRepresentation: TypeDisplayRepresentation = ""
    public var displayRepresentation: DisplayRepresentation { .init(stringLiteral: "") }
    public let id: String
}

// Trying 3 different strategies to use `@Dependency`
public struct AppEntityConformingModelQuery: EntityQuery {
    @Dependency(\.exampleDependency) var db // ERROR: Type of expression is ambiguous without a type annotation
    @Dependency(\.exampleDependency) var db: ExampleDependency // ERROR: Property type 'ExampleDependency' does not match 'wrappedValue' type '<<error type>>'

    public init() { 
        @Dependency(\.exampleDependency) var db // ERROR: Type of expression is ambiguous without a type annotation
    }
    public func entities(for identifiers: [AppEntityConformingModel.ID]) async throws -> [AppEntityConformingModel] { [] }
    public func suggestedEntities() async throws -> [AppEntityConformingModel] { [] }
    public func defaultResult() async -> AppEntityConformingModel? { nil }
}
```

For now, I think I can use the below workaround: using `@Dependency` in the `defaultQuery` property of the `AppEntity` and passing the results manually into the `AppEntityConformingModelQuery` initializer.

```swift
public struct AppEntityConformingModel: AppEntity {
    public static var defaultQuery: AppEntityConformingModelQuery {
        @Dependency(\.exampleDependency) var db
        return AppEntityConformingModelQuery(db: db)
    }
    // ...
}
```

Related docs for creating a configurable Widget that show minimal `EntityQuery` and `AppEntity` definitions: https://developer.apple.com/documentation/widgetkit/making-a-configurable-widget

### Dependencies version information

main (3815c48) (near 1.3.1)

### Destination operating system

iOS 17

### Xcode version information

Version 15.4 (15F31d)

### Swift Compiler version information

```shell
swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0
```

**Comment by mbrandonw** [comment]
Hi @twocentstudios, I believe this is because the AppIntents framework already has its own `@Dependency` property wrapper. Well, actually, they have a `@AppDependency` property wrapper, and then for some reason have a `typealias Dependency = AppDependency`, and there's not even any docs on it.

So unfortunately I believe this means you have to do:

```swift
@Dependencies.Dependency(\.key) var value
```

…when importing AppIntents.

I was under the impression that when there are conflicts between a user's code and Apple's libraries that you would be forced to qualify their APIs (e.g. `@AppIntents.Dependency`), but maybe that is only true of types in the standard library.

In any case, since this isn't an issue with the library I am going to convert it to a discussion. Please feel free to continue the conversation over there!

**Reply by twocentstudios** [reply]
Ah, that did the trick! Thanks 🙏

**Reply by Bartozo** [reply]
I was also wondering why it didn't work. Thank you so much for this solution 🙌

**Reply by Manuel-Kehl** [reply]
I was stuck on this going in circles for a while (and actually using questionable workarounds like a `sharedForAppIntents` Singleton "escape hatch" just so that I can access the objects I'd normally manage through Dependencies from AppIntents 😅)

Really glad I stumbled on this discussion as I can confirm that your proposed workaround to explicitly add the `Dependencies` prefix in my AppIntents code got it working as expected 🎉
Knowing about the conflicting `Dependencies typealias`, the compiler error makes a lot more sense to me now but it seemed a little cryptic and unexpected, when I first encountered it...

Since AppIntents are a pretty integral part of modern iOS apps, I'd expect many folks to run into this sooner or later, so I was wondering if it'd maybe make sense to just add short one-liner warning/note about this to the top level [README](https://github.com/pointfreeco/swift-dependencies/blob/main/README.md) file? Would probably reduce some friction to be aware of this before encountering the problem - even if not immediately applicable, it would be the kind of thing that'd make me go *"💡 Ah I remember reading about this in the README when I got started with Dependencies - let me quickly double-check"*.
